src/libs/vehicle/mavlink/vehicle.ts

```diff
@@ -773,6 +774,68 @@ export abstract class MAVLinkVehicle<Modes> extends Vehicle.AbstractVehicle<Mode
     sendMavlinkMessage(heartbeatMessage)
   }
 
+  /**
+   * Send system time from GCS
+   */
+  sendSystemTime(): void {
+    const systemTimeMessage: Message.SystemTime = {
+      type: MAVLinkType.SYSTEM_TIME,
+      time_unix_usec: Math.floor(performance.now() * 1000),
+      time_boot_ms: 0,
+    }
+    sendMavlinkMessage(systemTimeMessage)
+  }
+
+  /**
+   * Measure network latency to vehicle using TIMESYNC message
+   * @returns {Promise<void>}
+   */
+  async measureLatency(): Promise<void> {
+    const time_sent_ns = Math.floor(performance.now()) * 1e6
+    let receivedTimesync = false
+    let timeoutReached = false
+    let latencyNs = 0
+    const timeout = 5000
+
+    const timesyncHandler = (pack: Package): void => {
+      const timesync = pack.message as Message.Timesync
+      if (timesync.ts1 === time_sent_ns) {
+        const time_received_ns = Math.floor(performance.now()) * 1e6
+        // we could use the tc1 value to calculate the latency, but I think that would give us the
+        // GCS -> vehicle latency, not the round trip latency.
+        latencyNs = (time_received_ns - time_sent_ns) / 2
+        receivedTimesync = true
+      }
+    }
+
+    const timeSyncMessage: Message.Timesync = {
+      type: MAVLinkType.TIMESYNC,
+      ts1: time_sent_ns,
+      tc1: 0,
+      target_system: this.currentSystemId,
+      target_component: 1,
+    }
+    const startTime = new Date()
```

Comment:
This could also use `performance.now()`, although I'm not sure what's preferable (i.e. consistency of time functions within the file, vs consistency of `Date` used for timeouts throughout the codebase) 🤷‍♂️ 

Not fussed either way - just pointing out it's an option.